Deprecating CTE for Rails 7.2+ native support

ltrainpr opened the pull request:
FYI: I used Claude and ChatGPT agents to create this PR.

### New Features

- Added CTE deprecation support for Rails 7.2+
  - Configuration option to disable CTE functionality: `ActiveRecordExtended::CteDeprecationConfiguration.disable_cte_for_rails_7_2_plus = true`
- Configuration option to disable deprecation warnings: `ActiveRecordExtended::CteDeprecationConfiguration.cte_deprecation_warnings_enabled = false`
  - Deprecation warnings are shown when using CTE methods with Rails 7.2+
  - CTE functionality can be completely disabled for Rails 7.2+ to avoid conflicts with native Rails CTE support

GeorgeKaraszi commented on lib/active_record_extended/query_methods/with_cte.rb:
```diff
@@ -215,4 +235,11 @@ def add_materialized_modifier(expression, cte, name)
 
-ActiveRecord::Relation.prepend(ActiveRecordExtended::QueryMethods::WithCTE)
+# Only prepend the WithCTE module if CTE functionality is not disabled for Rails 7.2+
+unless ActiveRecordExtended::CteDeprecationConfiguration.should_disable_cte?
+  ActiveRecord::Relation.prepend(ActiveRecordExtended::QueryMethods::WithCTE)
+end
+# Only prepend the WithCTE module if CTE functionality is not disabled for Rails 7.2+
+unless ActiveRecordExtended::CteDeprecationConfiguration.should_disable_cte?
+  ActiveRecord::Relation.prepend(ActiveRecordExtended::QueryMethods::WithCTE)
+end
```
This will not work. You are going to run into the simple fact that this file is executed the moment the gem is loaded, giving no time to evaluate any modifications to the configuration.

GeorgeKaraszi commented on lib/active_record_extended/version.rb:
```diff
@@ -1,5 +1,5 @@
 
 module ActiveRecordExtended
-  VERSION = "3.3.0"
+  VERSION = "3.4.0"
```
Don't change the version

GeorgeKaraszi commented on lib/active_record_extended/patch/relation_patch.rb:
```diff
@@ -4,16 +4,25 @@ module ActiveRecordExtended
-        ).freeze
-
-        delegate(*AR_EX_QUERY_METHODS, to: :all)
+        BASE_QUERY_METHODS = [
+          :define_window, :select_window, :foster_select,
+          :either_join, :either_joins, :either_order, :either_orders
+        ].freeze
+
+        def self.query_methods
+          methods = BASE_QUERY_METHODS +
+                    ActiveRecordExtended::QueryMethods::Unionize::UNIONIZE_METHODS +
+                    ActiveRecordExtended::QueryMethods::Json::JSON_QUERY_METHODS
+
+          # Only include CTE methods if not disabled for Rails 7.2+
+          unless ActiveRecordExtended::CteDeprecationConfiguration.should_disable_cte?
+            methods << :with
+          end
+
+          methods
+        end
+
+        delegate(*query_methods, to: :all)
```
This will not work for the same reason this doesn't work for the `prepend` this is executed before any config can be modified.

GeorgeKaraszi commented on lib/active_record_extended/cte_deprecation_configuration.rb:
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module ActiveRecordExtended
+  class CteDeprecationConfiguration
```
I would rather change this into a module `Configuration`. Not a fan of this explicit deprecation module since it's too limiting.

GeorgeKaraszi commented on lib/active_record_extended/cte_deprecation_configuration.rb:
```diff
@@ -0,0 +1,42 @@
+      end
+
+      def rails_7_2_plus?
+        Gem::Requirement.new(">= 7.2").satisfied_by?(ActiveRecord.gem_version)
```
You already evaluated Rails 7.2 into a constant

GeorgeKaraszi reviewed (changes requested):
There should be a module Config that only deals with the boolean aspect. You should peek at other gems that implement configurations to draw some inspiration from.

If at all possible, let's not vibe code these changes. There's a lot of stuff that doesn't work here.

ltrainpr committed:
[LC-2396]: Deprecate and Feature-Flag WithCTE for Rails 7.2+

**Summary:**
- Adds a config-driven feature flag (`ActiveRecordExtended.config.with_cte_enabled`) to enable/disable `WithCTE` support at runtime.
- Emits deprecation warnings for `WithCTE` on Rails 7.2+ (native CTE support), with an option to silence warnings (`with_cte_deprecation_warnings_enabled`).
- Disabling the feature raises clear errors if `WithCTE` is used.
- Backward compatible: Rails < 7.2 is unaffected by default.
- Updated README and robust tests for all scenarios.
- Deprecator now uses the actual gem version.

**Migration:**
- For Rails 7.2+, add to an initializer to disable:
  ```ruby
  ActiveRecordExtended.config.with_cte_enabled = false
  ```
- To silence warnings:
  ```ruby
  ActiveRecordExtended.config.with_cte_deprecation_warnings_enabled = false
  ```

**Why:**
- Rails 7.2+ has native CTE support. This change allows safe, configurable migration for all users.

ltrainpr committed:
Update to not raise on rails native supported methods
